### PR TITLE
Adding support for CGC w/ Automation

### PR DESCRIPTION
**Pull request by jordan-rash**
CC: @adthrasher

### CODE REVIEW

File: Dockerfile.cgc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+RUN R -e 'BiocManager::install(c("biomaRt", "GSVA", "rhdf5", "ComplexHeatmap", "ConsensusClusterPlus", "DESeq2", "tximport", "impute", "limma", "GEOquery"))'
+
+# install NetBID2 from GitHub master branch
+RUN xvfb-run R -e 'devtools::install_github("jyyulab/NetBID", ref="master", dependencies="Depends")'
+
+WORKDIR /data
+COPY IBC_CCDI .
+
+ENTRYPOINT ["Rscript"]
+CMD ["IBC_CCDI.R"]
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline Dockerfile.cgc]:
```suggestion
CMD ["cloudAppNetBID.R"]
```

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      - 
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - 
+        name: Login to GitHub Container Registry
+        uses: docker/login-action@v1 
+        with:
+          registry: ghcr.io
+          username: ${{ secrets.CLOUDY_USERNAME }}
+          password: ${{ secrets.CLOUDY_TOKEN }}
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]: If Jiyang's group wants to deploy the container image to GitHub, they'll need to set the repo-level secrets (and probably will want to remove the CLOUDY reference).

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      -
+        name: Login to Cancer Genomics Cloud Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: cgc-images.sbgenomics.com
+          username: ${{ secrets.CGC_USERNAME }}
+          password: ${{ secrets.CGC_TOKEN }}
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml:39]: Jiyang's team will need to set the secrets on the repo to an account with the correct credentials to deploy to CGC.

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          context: .
+          file: ./Dockerfile.cgc
+          push: true
+          tags: |
+            ghcr.io/${{ env.REPO_LOWER }}:latest
+            ghcr.io/${{ env.REPO_LOWER }}:${{ env.SHORT_SHA }}
+            ghcr.io/${{ env.REPO_LOWER }}:${{ github.event.release.tag_name }}
+            cgc-images.sbgenomics.com/${{ env.REPO_LOWER }}:latest
+            cgc-images.sbgenomics.com/${{ env.REPO_LOWER }}:${{ env.SHORT_SHA }}
+            cgc-images.sbgenomics.com/${{ env.REPO_LOWER }}:${{ github.event.release.tag_name }}
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]: This assumes that the CGC username will be `jyyulab` and the project is `NetBID`. Jiyang's team will either need to create those exactly or update with the correct names.

File: cgc/netbid.cwl [hunk 5]
```diff
@@ -0,0 +1,101 @@
+{
+    "class": "CommandLineTool",
+    "cwlVersion": "v1.2",
+    "id": "cjrash/netbid/netbid/11",
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
```

File: cgc/netbid.cwl [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    "requirements": [
+        {
+            "class": "ShellCommandRequirement"
+        },
+        {
+            "class": "LoadListingRequirement"
+        },
+        {
+            "class": "DockerRequirement",
+            "dockerPull": "cgc-images.sbgenomics.com/cjrash/stjude/netbid:latest"
```
Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]: This can be updated once we know where Jiyang's team will deploy in CGC.

**Review by adthrasher — COMMENTED** [review]
We can remove the IBC files:
IBC_CCDI/TCGABRCA_log2fpkm.SIG.txt
IBC_CCDI/GEP.195.quantileNormalized.geneLevel_log2.eset
IBC_CCDI/TCGABRCA_log2fpkm.TF.txt

@adamdingliang - Someone from your team will need to draft a description for the tool in CGC. See this example: https://cgc.sbgenomics.com/public/apps#ZCHONG/novobreak-commit/novobreak-analysis/. It would also be best to have helpful documentation for each of the input to the script (expression set, transcription factor network, signaling network, and metadata file). This should include information about the file formats and how to generate them. Ideally, it would also link to examples.

File: .github/workflows/publish.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (adthrasher) [reply .github/workflows/publish.yml:39]: CGC_TOKEN and CGC_USERNAME has been added

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [reply .github/workflows/publish.yml:39]: CGC_TOKEN and CGC_USERNAME has been added

File: .github/workflows/publish.yml (hunk 2 above)
Comment (adthrasher) [inline .github/workflows/publish.yml:32]:
```suggestion
```
Since NetBID has a separate Docker image, we will just build and deploy to CGC.

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]:
```suggestion
```
Since NetBID has a separate Docker image, we will just build and deploy to CGC.

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./Dockerfile.cgc
+          push: true
+          tags: |
+            ghcr.io/${{ env.REPO_LOWER }}:latest
+            ghcr.io/${{ env.REPO_LOWER }}:${{ env.SHORT_SHA }}
+            ghcr.io/${{ env.REPO_LOWER }}:${{ github.event.release.tag_name }}
```
Comment (adthrasher) [inline .github/workflows/publish.yml:51]:
```suggestion
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml:51]:
```suggestion
```

File: .github/workflows/publish.yml (hunk 4 above)
Comment (adthrasher) [inline .github/workflows/publish.yml:54]:
```suggestion
            cgc-images.sbgenomics.com/stjude/netbid:latest
            cgc-images.sbgenomics.com/stjude/netbid:${{ env.SHORT_SHA }}
            cgc-images.sbgenomics.com/stjude/netbid:${{ github.event.release.tag_name }}
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]:
```suggestion
            cgc-images.sbgenomics.com/stjude/netbid:latest
            cgc-images.sbgenomics.com/stjude/netbid:${{ env.SHORT_SHA }}
            cgc-images.sbgenomics.com/stjude/netbid:${{ github.event.release.tag_name }}
```

File: cgc/netbid.cwl [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        },
+        {
+            "id": "netbid_log",
+            "type": "stdout",
+            "outputBinding": {
+                "glob": "$(inputs.expression_set.nameroot).log"
+            }
+        }
+    ],
```
Comment (adthrasher) [inline cgc/netbid.cwl:81]:
```suggestion
    ],
    "doc": "# Description\n\nNetBID is a data-driven system biology pipeline using a data-driven network-based Bayesian inference approach to find drivers from transcriptomics, proteomics, or phosphoproteomics data. The drivers can be either transcription factors (TF) or signaling factors (SIG).\n\nNetBID2 has the following key steps to perform hidden driver analysis:\n1.\tActivity calculation of drivers based on drivers’ regulons from a pre-built or user-provided SJARACNe network;\n2.\tDiscovery of differential expressed genes and differential activated drivers;\n3.\tGeneration of the master table for drivers;\n4.\tVisualizing drivers with significance profiles and target genes.\n\n# Inputs and outputs of NetBID workflow\n## Inputs:\n*\tExpression matrix - comma-delimited expression matrix file with columns as samples, rows as genes.\n*\tMetadata file - comma-delimited file with two columns: sample and group.\n*\tSignaling (SIG) network - file with each row an edge from the SIG network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)\n*\tTranscription factor (TF) network - file with each row an advantage from the TF network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)\n\n## Outputs:\n*\tExcel file with differential expressed genes and differential activated drivers\n*\tPicture file visualizing drivers with significance profiles\n\n# Common issues\n*\tThe first row and the first column of the expression matrix file must be sample names and gene names, respectively.\n*\tThe metadata file must have at least two sample groups in the 2nd column.",

```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl:80]:
```suggestion
    ],
    "doc": "# Description\n\nNetBID is a data-driven system biology pipeline using a data-driven network-based Bayesian inference approach to find drivers from transcriptomics, proteomics, or phosphoproteomics data. The drivers can be either transcription factors (TF) or signaling factors (SIG).\n\nNetBID2 has the following key steps to perform hidden driver analysis:\n1.\tActivity calculation of drivers based on drivers’ regulons from a pre-built or user-provided SJARACNe network;\n2.\tDiscovery of differential expressed genes and differential activated drivers;\n3.\tGeneration of the master table for drivers;\n4.\tVisualizing drivers with significance profiles and target genes.\n\n# Inputs and outputs of NetBID workflow\n## Inputs:\n*\tExpression matrix - comma-delimited expression matrix file with columns as samples, rows as genes.\n*\tMetadata file - comma-delimited file with two columns: sample and group.\n*\tSignaling (SIG) network - file with each row an edge from the SIG network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)\n*\tTranscription factor (TF) network - file with each row an edge from the TF network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)\n\n## Outputs:\n*\tExcel file with differential expressed genes and differential activated drivers\n*\tPicture file visualizing drivers with significance profiles\n\n# Common issues\n*\tThe first row and the first column of the expression matrix file must be sample names and gene names, respectively.\n*\tThe metadata file must have at least two sample groups in the 2nd column.",

```

File: cgc/netbid.cwl [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "id": "expression_set",
+            "type": "File",
+            "inputBinding": {
+                "prefix": "-e",
+                "shellQuote": false,
+                "position": 1
+            },
+            "label": "Gene Expression Matrix"
```
Comment (adthrasher) [inline cgc/netbid.cwl:22]:
```suggestion
            "label": "Gene Expression Matrix",
            "doc": "comma-delimited expression matrix file with columns as samples, rows as genes."

```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
            "label": "Gene Expression Matrix",
            "doc": "comma-delimited expression matrix file with columns as samples, rows as genes."

```

File: cgc/netbid.cwl [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "id": "tf",
+            "type": "File",
+            "inputBinding": {
+                "prefix": "-t",
+                "shellQuote": false,
+                "position": 1
+            },
+            "label": "Transcription Factor Network"
```
Comment (adthrasher) [inline cgc/netbid.cwl:32]:
```suggestion
            "label": "Transcription Factor Network",
            "doc": "file with each row an edge from the TF network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)"

```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
            "label": "Transcription Factor Network",
            "doc": "file with each row an edge from the TF network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)"

```

File: cgc/netbid.cwl [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "id": "sig",
+            "type": "File",
+            "inputBinding": {
+                "prefix": "-s",
+                "shellQuote": false,
+                "position": 1
+            },
+            "label": "Signaling Network"
```
Comment (adthrasher) [inline cgc/netbid.cwl:42]:
```suggestion
            "label": "Signaling Network",
            "doc": "file with each row an edge from the SIG network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)"

```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
            "label": "Signaling Network",
            "doc": "file with each row an edge from the SIG network constructed using SJARACNe (https://github.com/jyyulab/SJARACNe)"

```

File: cgc/netbid.cwl [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "id": "metadata",
+            "type": "File",
+            "inputBinding": {
+                "prefix": "-m",
+                "shellQuote": false,
+                "position": 3
+            },
+            "label": "Sample Grouping File"
```
Comment (adthrasher) [inline cgc/netbid.cwl:52]:
```suggestion
            "label": "Sample Grouping File",
            "doc": "comma-delimited file with two columns: sample and group."
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
            "label": "Sample Grouping File",
            "doc": "comma-delimited file with two columns: sample and group."
```

File: cgc/netbid.cwl [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    "baseCommand": [],
+    "inputs": [
+        {
+            "id": "rscript",
+            "type": "File",
+            "inputBinding": {
+                "shellQuote": false,
+                "position": 0
+            }
+        },
```
Comment (adthrasher) [inline cgc/netbid.cwl:13]:
```suggestion
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
```

File: cgc/netbid.cwl [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    "requirements": [
+        {
+            "class": "ShellCommandRequirement"
+        },
+        {
+            "class": "LoadListingRequirement"
+        },
+        {
+            "class": "DockerRequirement",
+            "dockerPull": "cgc-images.sbgenomics.com/cjrash/stjude/netbid:latest"
```
Comment (adthrasher) [inline cgc/netbid.cwl:92]:
```suggestion
            "dockerPull": "cgc-images.sbgenomics.com/stjude/netbid:latest"
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline cgc/netbid.cwl]:
```suggestion
            "dockerPull": "cgc-images.sbgenomics.com/stjude/netbid:latest"
```

File: .github/workflows/publish.yml [hunk 15]
```diff
@@ -23,13 +23,6 @@ jobs:
       - 
         name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v1
-      - 
-        name: Login to GitHub Container Registry
-        uses: docker/login-action@v1 
-        with:
-          registry: ghcr.io
-          username: ${{ secrets.CLOUDY_USERNAME }}
-          password: ${{ secrets.CLOUDY_TOKEN }}
```
Comment (adthrasher) [inline .github/workflows/publish.yml]: From yesterday's discussion, we are going to need to revert this change so that we can push the image to ghcr. We should update the secret names to `GH_USERNAME` and `GH_TOKEN`. I've asked Adam to create those secrets on the repo.

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]: From yesterday's discussion, we are going to need to revert this change so that we can push the image to ghcr. We should update the secret names to `GH_USERNAME` and `GH_TOKEN`. I've asked Adam to create those secrets on the repo.

File: .github/workflows/publish.yml [hunk 16]
```diff
@@ -39,9 +39,6 @@ jobs:
           file: ./Dockerfile.cgc
           push: true
           tags: |
-            ghcr.io/${{ env.REPO_LOWER }}:latest
-            ghcr.io/${{ env.REPO_LOWER }}:${{ env.SHORT_SHA }}
-            ghcr.io/${{ env.REPO_LOWER }}:${{ github.event.release.tag_name }}
```
Comment (adthrasher) [inline .github/workflows/publish.yml]: We'll need to revert this change also after yesterday's discussion.

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]: We'll need to revert this change also after yesterday's discussion.

File: .github/workflows/publish.yml (hunk 15 above)
Comment (adthrasher) [reply .github/workflows/publish.yml]: @jordan-rash

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [reply .github/workflows/publish.yml]: @jordan-rash

File: .github/workflows/publish.yml (hunk 16 above)
Comment (adthrasher) [reply .github/workflows/publish.yml]: @jordan-rash

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [reply .github/workflows/publish.yml]: @jordan-rash

**Comment by jordan-rash** [comment]
@adthrasher I just rebased and squashed the commits while dropping the GH change.....I am reviewing now, but another set of eyes is helpful....when I use the "Commit Changes" button in the UI, the commit log is not helpful, so I might have missed a piece

File: .github/workflows/publish.yml (hunk 2 above)
Comment (adthrasher) [inline .github/workflows/publish.yml:32]:
```suggestion
          username: ${{ secrets.GH_USERNAME }}
          password: ${{ secrets.GH_TOKEN }}
```

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [inline .github/workflows/publish.yml]:
```suggestion
          username: ${{ secrets.GH_USERNAME }}
          password: ${{ secrets.GH_TOKEN }}
```

**Review by adthrasher — APPROVED** [review]
Just the small change to the environment secrets, then it looks good.

File: .github/workflows/cgc.yml [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+      - uses: actions/checkout@v2
+      - name: Set the docker tag from Repo Tag
+        id: set_dockertag
+        env:
+          IMAGE: cgc-images.sbgenomics.com/stjude/netbid
+          VERSION_TAG: ${{ github.event.release.tag_name }}
+        run: |
+          jq --arg image "${{ env.IMAGE }}:${{ env.VERSION_TAG }}" '(.requirements | .[] | select(.class == ("DockerRequirement")) | .dockerPull) |= $image' cgc/netbid.cwl > cgc/netbid.cwl.new
+          mv cgc/netbid.cwl.new cgc/netbid.cwl
+          cat cgc/netbid.cwl
```
Comment (jordan-rash) [inline .github/workflows/cgc.yml:21]: @adthrasher Ran this locally via `act` and it looks to work okay.  let me know what you think

Comment (Tgu6Qn6P7rhOT294U1BtXU306jB5Yucd6eqGLsG) [inline .github/workflows/cgc.yml:21]: @adthrasher Ran this locally via `act` and it looks to work okay.  let me know what you think

Comment (adthrasher) [reply .github/workflows/cgc.yml:21]: I think that looks good.

Comment (trc9VDDnTMJlxo9PomkjwRHTB9rcqDUwUWW1THp) [reply .github/workflows/cgc.yml:21]: I think that looks good.

**Review by adthrasher — APPROVED** [review]

**Review by adamdingliang — APPROVED** [review]
Completed sanity check
